[CR]Make map::nul* fields static

Coolthulhu opened the pull request:
From #12314

```
|error: use of deleted function 'map& map::operator=(map&&)'|
|note: 'map& map::operator=(map&&)' is implicitly deleted because the default definition would be ill-formed:|
|error: non-static data member 'map::nulfield' does not have a move assignment operator or trivial copy assignment operator|
|error: non-static data member 'map::nulveh' does not have a move assignment operator or trivial copy assignment operator|
|error: non-static data member 'map::nullcache' does not have a move assignment operator or trivial copy assignment operator|
```

Changed the `map::nul*` fields to static and defined them in map.cpp. That will have an additional advantage of making tinymaps a tiny bit more tiny.

Not sure if I'm not breaking some standard here. From what I've seen from a quick search it is fine, but there may be some compiler issue I'm missing. Compiles and works on g++ 4.9.1

BevapDin commented on src/map.cpp:
```diff
@@ -40,6 +40,13 @@ extern bool is_valid_in_w_terrain(int,int);
+std::list<item>  map::nulitems;          // Returned when &i_at() is asked for an OOB value
+ter_id           map::nulter;            // Returned when &ter() is asked for an OOB value
+field            map::nulfield;          // Returned when &field_at() is asked for an OOB value
+vehicle          map::nulveh;            // Returned when &veh_at() is asked for an OOB value
```
Is this one actually used anywhere? Vehicles are usually returned as pointers / nullptr, so there is no need for a dummy null-object.

Coolthulhu replied:
Never seen it used anywhere. Looks like someone wanted to add a `&veh_at`, but never got to it.

BevapDin replied:
What I meant: if it is not used, why not remove it? It can always be added back, but keeping things in the code that are not used at all is only confusing and slows down compilation. It's the same reasoning for functions that are not used.

Coolthulhu commented:
OK, removed it and updated for master.

By the way, I recall someone saying that tinymap has a lot of caches. What extra caches do maps carry now? Is there a good way to check all the actually produced members of a class (preferably with a `sizeof` for every one of them)?

DavidKeaton commented:
tbh I was more concerned with getting the fixes out, so AIM wasn't broken anymore. gonna make my PRs smaller now, like you suggested. :-P

BevapDin commented:
> I recall someone saying that tinymap has a lot of caches

It inherits *all* the caches from `map`. The inheritance is kind of wrong. One would expect a `map` to inherit from `tinymap` because it adds data and thereby becomes an "grown up" (== not tiny) map.

> Is there a good way to check all the actually produced members of a class (preferably with a  sizeof  for every one of them)?

No idea, maybe the debugger provides that info? Otherwise just create a string at runtime and show it in a debug window or so.

BevapDin commented:
It seems you can also remove `nulter`, it's assigned but never read anywhere.

There is a `null_submap` defined in map.cpp, which reminds me: you could remove the declaration of those null objects from the header. They don't need to be exposed, they can exist as static object inside of map.cpp.

And this might actually allow to remove vehicle.h from map.h and from mapdata.h, so it is not included in any header anymore.

I'm sorry for giving you more work. I can merge this and add those changes later myself, do you prefer that?

Coolthulhu commented:
I'll add them from my side. If I miss something, you can add it on your side or send as a PR - I'll be mergetesting after you're done.

Coolthulhu commented:
Removal of vehicle.h from mapdata.h cascaded a bit - a lot of files have line.h only added through vehicle.h and this big untangle makes them complain.

Coolthulhu commented:
Are you working on the mapdata.h part? I could finish it from my side and send as a separate PR, but if you're halfway there then I'll wait.

Apparently many files only include vehicle.h through mapdata.h.

BevapDin commented:
> Apparently many files only include vehicle.h through mapdata.h.

Yes, and many don't even need it, they need line.h sometimes. See #12438, compiling toke some time.
